api/types/instance_test.go

```diff
@@ -23,6 +23,142 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestInstanceFilter(t *testing.T) {
+	iis := []struct {
+		id       string
+		version  string
+		services []SystemRole
+		upgrader string
+	}{
+		{
+			id:       "a1",
+			version:  "v1.2.3",
+			services: []SystemRole{RoleAuth},
+		},
+		{
+			id:       "a2",
+			version:  "v2.3.4",
+			services: []SystemRole{RoleAuth, RoleNode},
+			upgrader: "kube",
+		},
+		{
+			id:       "p1",
+			version:  "v1.2.1",
+			services: []SystemRole{RoleProxy},
+		},
+		{
+			id:       "p2",
+			version:  "v2.3.1",
+			services: []SystemRole{RoleProxy, RoleNode},
+			upgrader: "unit",
+		},
+	}
+
+	// set up group of test instances
+	var instances []Instance
+	for _, ii := range iis {
+		ins, err := NewInstance(ii.id, InstanceSpecV1{
+			Version:          ii.version,
+			Services:         ii.services,
+			ExternalUpgrader: ii.upgrader,
+		})
+
+		require.NoError(t, err)
```

Comment:
nit: include `ii.id` in the error message, just so that we have more than the line number to go on if this fails.
```suggestion
		require.NoError(t, err, "id: %v", ii.id)
```